idseq_pipeline/commands/non_host_alignment_functions.py

```diff
@@ -677,28 +590,39 @@ def interpret_min_column_number_string(min_column_number_string, correct_number_
         min_column_number = correct_number_of_output_columns
     return min_column_number
 
+def add_blank_line(file_path):
+    execute_command("echo '' >> %s;" % file_path)
+
+def remove_blank_line(file_path):
+    execute_command("sed -i '$ {/^$/d;}' %s" % file_path)
 
 def run_gsnapl_chunk(part_suffix, remote_home_dir, remote_index_dir, remote_work_dir, remote_username,
                      input_files, key_path, lazy_run):
     chunk_id = input_files[0].split(part_suffix)[-1]
-    outfile_basename = 'gsnapl-out' + part_suffix + chunk_id
-    dedup_outfile_basename = 'dedup-' + outfile_basename
-    local_outfile = CHUNKS_RESULT_DIR + "/" + outfile_basename
-    remote_outfile = os.path.join(remote_work_dir, outfile_basename)
     commands = "mkdir -p %s;" % remote_work_dir
     for input_fa in input_files:
         commands += "aws s3 cp --quiet %s/%s %s/ ; " % \
                  (SAMPLE_S3_OUTPUT_CHUNKS_PATH, input_fa, remote_work_dir)
+    multihit_basename = "multihit-gsnapl-out" + part_suffix + chunk_id
+    multihit_local_outfile = os.path.join(CHUNKS_RESULT_DIR, multihit_basename)
+    multihit_remote_outfile = os.path.join(remote_work_dir, multihit_basename)
     commands += " ".join([remote_home_dir+'/bin/gsnapl',
                           '-A', 'm8', '--batch=0', '--use-shared-memory=0',
-                          '--gmap-mode=none', '--npaths=1', '--ordered',
+                          '--gmap-mode=none', '--npaths=1000', '--ordered',
                           '-t', '32',
-                          '--maxsearch=5', '--max-mismatches=40',
+                          '--maxsearch=1000', '--max-mismatches=40',
                           '-D', remote_index_dir, '-d', 'nt_k16']
                          + [remote_work_dir+'/'+input_fa for input_fa in input_files]
-                         + ['> '+remote_outfile, ';'])
-
-    if not lazy_run or not fetch_lazy_result(os.path.join(SAMPLE_S3_OUTPUT_CHUNKS_PATH, dedup_outfile_basename), CHUNKS_RESULT_DIR):
+                         + ['> ' + multihit_remote_outfile, ';'])
+
+    multihit_summary_basename = "summary-" + multihit_basename
+    dedup_multihit_basename = "dedup-" + multihit_basename
+    multihit_summary_file = os.path.join(CHUNKS_RESULT_DIR, multihit_summary_basename)
+    dedup_multihit_local_outfile = os.path.join(CHUNKS_RESULT_DIR, dedup_multihit_basename)
+    is_multihit_present = fetch_lazy_result(os.path.join(SAMPLE_S3_OUTPUT_CHUNKS_PATH, multihit_basename), CHUNKS_RESULT_DIR)
+    is_multihit_summary_present = fetch_lazy_result(os.path.join(SAMPLE_S3_OUTPUT_CHUNKS_PATH, multihit_summary_basename), CHUNKS_RESULT_DIR)
+    is_dedup_multihit_present = fetch_lazy_result(os.path.join(SAMPLE_S3_OUTPUT_CHUNKS_PATH, dedup_multihit_basename), CHUNKS_RESULT_DIR)
+    if not lazy_run or not is_multihit_present or not is_multihit_summary_present or not is_dedup_multihit_present:
```

Comment:
Since we are always lazy this here makes no difference, but in general, in the eager case, it's an unnecessary penalty in performance to try to fetch_lazy_result if we are then going to overwrite it.

Reply:
The way this code was written previously, it did not attempt the fetch unless lazy_run=True.

Reply:
In fact, if any one of these files is missing, there is no benefit to copying the other from s3.

Reply:
So, you should not break it down on separate lines like this.

Instead, write a little helper function here

```
    def lazy_fetch(basename):
        return fetch_lazy_result(os.path.join(SAMPLE_S3_OUTPUT_CHUNKS_PATH, basename), CHUNK_RESULTS_DIR)
```
then you can write the if condition as follows
```
    if not lazy_run or not all(lazy_fetch(f) for f in [multihit_basename, multihit_summary_basename, dedup_multihit_basename]):
        ...
```

Reply:
In fact for readability you may want to make to call the helper function `lazy_fetch_all` so it gets a list
```
    def lazy_fetch_all(basenames):
        return all(fetch_lazy_result(os.path.join(SAMPLE_S3_OUTPUT_CHUNKS_PATH, bn), CHUNK_RESULTS_DIR) for bn in basenames)
```
then the if guard would read
```
    if not lazy_run or not lazy_fetch_all([multihit_basename, multihit_summary_basename, dedup_multihit_basename]):
        ...
```

Reply:
ok